private/buf/buflsp/file.go

```diff
@@ -736,6 +743,41 @@ func findImportable(
 	return imports, nil
 }
 
+// getWorkspaceWithLocalPaths checks each proto file in a workspace and ensures they all have
+// a LocalPath. If not, then it refreshes the workspace. As outlined on [storage.ObjectInfo],
+// LocalPath will not always be present if the file did not originate from disk. If an import's
+// module data was fetched remotely, then it will not have a path on disk to the cache. By
+// refreshing the module, and rebuilding the workspace, we populate the module data from the cache.
+func getWorkspaceWithLocalPaths(ctx context.Context, fileName string, lsp *lsp) (bufworkspace.Workspace, error) {
+	workspace, err := lsp.controller.GetWorkspace(ctx, fileName)
+	if err != nil {
+		return nil, err
+	}
+	var refreshWorkspace bool
+	for _, module := range workspace.Modules() {
```

Comment:
This is very expensive - this means that for every call of `getWorkspaceWithLocalPaths`, you are going to iterate over every file in the workspace. This feels like it would be prohibitive.

Reply:
Yes -- covered some of this in the response to a comment below. If we do pursue changing the semantics `ModuleDataProvider` to return file info after writing to the cache, then we can avoid checking each file since we'll have guarantees to the file on-disk in the cache.